### PR TITLE
Adds support for takeoff

### PR DESCRIPTION
**Pull request by arjo129**
This commit adds support for takeoff behaviour in the flight controller. I also discovered that the pitch axis was flipped. This PR fixes it as well. It also removes the dependency on ros within the ignition plugin. A separate ros bridge was added in #81. The aim is to move that to another bridge node.

takeoff

## To test
You will need 3 terminals. In one terminal start gazebo.
```
ign gazebo coast.sdf
```
In another spawn the plane:
```
# Spawns on runway.
ros2 launch mbzirc_ign spawn.launch.py name:=zephyr world:=coast model:=mbzirc_fixed_wing type:=uav x:=-1500 y:=0 z:=5
```
Finally command the plane to take off:
```
ign service -s /zephyr/cmd/takeoff  --reqtype ignition.msgs.Float_V --reptype ignition.msgs.Boolean -r "data: [0.2, 40]" --timeout 10000000
```
The plane should start flying. The format of the last service is [intended pitch, target altitude]
## Todo
- [x] updates tests
- [x] update mavros command interface (see #81)
- [ ] write docs

Signed-off-by: Arjo Chakravarty <arjo@openrobotics.org>

### CODE REVIEW

File: mbzirc_ign/src/FixedWingController.cc [hunk 1]
```diff
@@ -83,67 +106,125 @@ class mbzirc::FixedWingControllerPrivate
...
-
-  /// \brief Attitude subscriber
-  public: rclcpp::Subscription<mavros_msgs::msg::AttitudeTarget>::SharedPtr
-    attitudeTargetSub;
-
   /// \brief Setup the ros node
   /// \param[in] _name Ros node name
   /// \param[in] _topic Topic to listen on
-  public: void SetupROS(const std::string &_name, const std::string &_topic)
+  public: void SetupROS(std::string _name, std::string _topic)
```
Comment (iche033) [inline mbzirc_ign/src/FixedWingController.cc:127]: change name and doxygen as it is no longer setting up ROS node / subcriber

File: mbzirc_ign/src/FixedWingController.cc [hunk 2]
```diff
@@ -83,67 +106,125 @@ class mbzirc::FixedWingControllerPrivate
...
     std::lock_guard<std::mutex> lock(this->mutex);
     math::Vector3d euler = quat.Euler();
     targetRoll = euler.X();
     targetPitch = euler.Y();
-    targetVelocity = _msg->thrust;
+    targetVelocity = _msg.data(4);
+  }
+
+  /// \brief Take off service
+  /// \param[in]
```
Comment (iche033) [inline mbzirc_ign/src/FixedWingController.cc:158]: add doc for params

File: mbzirc_ign/src/FixedWingController.cc [hunk 3]
```diff
@@ -83,67 +106,125 @@ class mbzirc::FixedWingControllerPrivate
...
+      _result.set_data(false);
+      return false;
+    }
+    this->takeOffMinPitch = takeoff_params.data(0);
+    this->takeOffAltitude = takeoff_params.data(1);
+    this->mode = Mode::TAKE_OFF;
+    _result.set_data(true);
+
+    //std::unique_lock lk(takeOffMutex);
+    //takeOffCv.wait(lk, [this]{return takeOffFlag;});
```
Comment (iche033) [inline mbzirc_ign/src/FixedWingController.cc:176]: remove?

File: mbzirc_ign/src/FixedWingController.cc [hunk 4]
```diff
@@ -231,59 +314,100 @@ void FixedWingControllerPlugin::Configure(
...
+
+      /// Once target altitude is reached level out.
+      if (pose->Data().Z() >= this->dataPtr->takeOffAltitude) {
+        this->dataPtr->mode = FixedWingControllerPrivate::Mode::ATTITUDE_CONTROL;
+        this->dataPtr->targetPitch = 0;
+        //{
+        //  std::lock_guard<std::mutex> lock(this->dataPtr->takeOffMutex);
+        //  this->dataPtr->takeOffFlag = true;
+        //  this->dataPtr->takeOffCv.notify_one();
+        //}
```
Comment (iche033) [inline mbzirc_ign/src/FixedWingController.cc:382]: remove?

File: mbzirc_ign/test/test_fixed_wing.cc [hunk 5]
```diff
@@ -77,12 +54,18 @@ TEST_F(MBZIRCTestFixture, FixedWingController)
    // Create a Gazebo world
   LoadWorld("empty_platform.sdf");
 
-  std::atomic<bool> keepRosRunning {true};
+  std::atomic<bool> keepRunning {true};
   auto rosThread = std::thread([&](){
```
Comment (iche033) [inline mbzirc_ign/test/test_fixed_wing.cc:58]: update name of thread var?

File: mbzirc_ign/CMakeLists.txt [hunk 6]
```diff
@@ -119,8 +116,8 @@ if(BUILD_TESTING)
       test/${TEST_TARGET}.cc
     )
     target_include_directories(${TEST_TARGET}
-      PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_BINARY_DIR}/proto mavros_msgs)
-    target_link_libraries(${TEST_TARGET} ignition-gazebo${IGN_GAZEBO_VER}::ignition-gazebo${IGN_GAZEBO_VER} rclcpp::rclcpp ${mavros_msgs_LIBRARIES})
+      PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_BINARY_DIR}/proto)
+    target_link_libraries(${TEST_TARGET} ignition-gazebo${IGN_GAZEBO_VER}::ignition-gazebo${IGN_GAZEBO_VER} rclcpp::rclcpp)
```
Comment (iche033) [inline mbzirc_ign/CMakeLists.txt:120]: remove `rclcpp::rclcpp`?

File: mbzirc_ign/package.xml [hunk 7]
```diff
@@ -16,7 +16,6 @@
   <build_depend>rclcpp</build_depend>
```
Comment (iche033) [inline mbzirc_ign/package.xml:16]: remove `rclcpp`?

**Review by iche033 — COMMENTED** [review]
nice! works well for me

**Comment by iche033** [comment]
On thing I just noticed is that if I leave the UAV on the run way for a while, e.g. 1 min sim time, then invoke the take-off command, the UAV seems to pitch down more and has a hard time taking off

**Comment by arjo129** [comment]
@iche033 I think 9d2dcd7 might resolve it. I reallized I was not resetting the PID between state transitions hence the I term was building up.

**Comment by iche033** [comment]
yep [that re](https://github.com/osrf/mbzirc/commit/9d2dcd79e937ee8e7436aaff3256cbe653c44111) seems to have fixed the issue for me.

While testing, I noticed that that the fixed wing  UAV slowly slides forward because of the wind I think. I temporarily added a thin block in the middle of the start gate, which acts like a wheel chock for the drone so that it still has runway left to take off.

**Review by iche033 — APPROVED** [review]

**Pull request merged**